jupyter_client/manager.py

```diff
@@ -337,10 +337,12 @@ def finish_shutdown(self, waittime=None, pollinterval=0.1):
                 self.log.debug("Kernel is taking too long to finish, killing")
                 self._kill_kernel()
 
-    def cleanup(self, connection_file=True):
+    def cleanup(self, restart=False):
         """Clean up resources when the kernel is shut down"""
-        if connection_file:
+        if not restart:
             self.cleanup_connection_file()
+            # shutdown ZMQ context as we no longer using this kernel
+            self.context.destroy(linger=100)
```

Comment:
It is not safe to assume that the context is owned by this KernelManager. It can (and likely should be) shared across all kernel managers in a given process, so destroying it unconditionally is probably not ideal.

Reply:
I thought we had switched the contexts to be 1:1 with managers because they're not threadsafe. But it's been a while since we made changes around that.

Reply:
contexts are threadsafe, sockets are not. There was an [issue](https://github.com/jupyter/jupyter_client/pull/437) around *multiprocessing* which I believe is also fixed in recent-enough pyzmq (if not, it's a pyzmq bug).

Reply:
I also just noticed the `_close_control_socket` below - destroying the context should occur after closing sockets.

Reply:
> I also just noticed the _close_control_socket below - destroying the context should occur after closing sockets.

That is a good catch

Reply:
Aha, nice catch. I missed this during refactoring. Let me fix this in new revision